common/src/main/java/io/netty/util/concurrent/DefaultPromise.java

```diff
@@ -420,18 +432,88 @@ private void notifyListeners() {
         }
         EventExecutor executor = executor();
         if (executor.inEventLoop()) {
```

Comment:
> The transport focused executors (backed by SingleThreadEventExecutor) could also get into trouble.

I don't believe that is true. `SingleThreadEventExecutor` uses a queue. Same with `EmbeddedEventLoop`. Any executor that uses a queue should be fine with the approach I mentioned (still use ThreadLocal, but when too large schedule a Runnable).

> We may be able to do a similar type of fix in ImmediateExecutor, 'EmbeddedEventLoop', 'ImmediateEventExecutor, transport executors, etc... in Netty ... but there are also other implementations ofExecutorthat we have no control over (things similar to guava'sDirectExecutor`).

I only believe `ImmediateEventExecutor` and `directExecutor` are impacted. You're talking about working around the stack depth problem for every _user_ of an Executor (which there are many more of than Executor implementations). And it seems using `ImmediateEventExecutor` would be rare for Promise given Netty's synchronization model.

It may be a different conversation if we were talking about ListenableFuture in Guava, because directExecutor is used _very_ frequently there since each callback provides its own independent Executor. And nothing is able to optimize for `inEventLoop()` which causes greater use of directExecutor. And you can't call ctx.executor() in code outside of Netty, so if you need an executor it has to be injected.